### ISSUE TITLE
World generation error feedback

### ISSUE THREAD

**Issue by Frisk202020** [ArchipelagoMW/Archipelago#6418]
### What feature would you like to see?

I've spent a week trying to debug the generation of my world because many locations failed the accessibility check even if all items required were seemingly accessible (looking sphere-by-sphere on a fixed seed). 
The issue ? Turns out some items were marked as `useful` instead of `progression`. I'll admit it's a skill issue on my part, but still, getting a generic FillError was really not helpful, quite the opposite in fact. This error rather lead me to inspect again and again region connection errors, which in fact were non-existent.

Maybe I'm missing a point, but I don't see a valid situation were you would have non-progression items appear into the rules. So I'm thinking, would it be possible to have a *warning* upon generation if such situation is observed ?

I know it's not straightforward to implement, indeed rules are registered as, ultimately, a lambda function as I understood the implementation : there's currently no way to go back from the rule to the items. Maybe we could just raise the warning if a non-progression item is used with a `Has / HasAll` ?

**Comment by Zunawe** [comment]
Only item _instances_ can be progression. Item names (which are all we really get in rules) don't indicate on their own whether one or more instances of that item that the player can have in their inventory is progression. So the seemingly trivial solution doesn't work.

That said, this is the sort of thing that would be nice to catch, so if there's a cleverer solution to detecting it, that would be neat. Or a sensible place to suggest checking this for devs who might have this problem.

**Comment by Mysteryem** [comment]
> Maybe I'm missing a point, but I don't see a valid situation were you would have non-progression items appear into the rules. So I'm thinking, would it be possible to have a warning upon generation if such situation is observed ?

> Maybe we could just raise the warning if a non-progression item is used with a `Has / HasAll` ?

Some worlds make only some copies progression, or (while it would be poor form for Rule Builder rules that should statically get rid of such checks) deliberately remove progression classification from items to disable them from logic, e.g. to disable logic from expecting certain tricks, to make the logic easier.

There are also worlds that override `collect_item()` so that collecting an item collects a different item into the state (generally used for progressive items where there is a toggle for progressive/non-progressive items, so that the logic rules don't need to care whether progressive/non-progressive is enabled), and worlds that override `collect()` to collect additional 'fake' items into the state, so some logic rules may never check for real items being in a state.

I have been writing a fuzzer hook (no public release yet) to detect cases where an item not being progression classification changed the logical spheres of the multiworld, but as outlined above, you cannot directly infer that failing this fuzzer hook is a bug in the world being fuzzed, and it currently only checks rules against items that exist in the multiworld (so cannot detect issues with `collect_item()`/`collect()` use).

**Comment by PoryGone** [comment]
Additionally, it's extremely common to have items where some of the instances are `progression` and the rest are `useful` or `filler`. This is not really an easily-solvable problem.

**Comment by Mysteryem** [comment]
The #ap-world-dev channel in the Archipelago discord server is a place you can ask questions about apworld development and issues you may come across while developing an apworld.
